StainedGlassLamp

echadwick-wayfair opened the pull request:
An example model to test and demonstrate KHR_materials_transmission, KHR_texture_clearcoat, KHR_materials_variants, KHR_texture_basisu, KHR_materials_translucency, KHR_materials_ior, and KHR_materials_volume.

Edit: adding a sample image of the asset.
![screenshot_large](https://user-images.githubusercontent.com/40177036/105220206-6452d600-5b25-11eb-96ee-c5d71e437222.jpg)

cx20 commented:
Hmm, I don't know why, but when I tried to download the following repository with zip, I got a network error at `1GB` and could not download.  
https://github.com/echadwick-wayfair/glTF-Sample-Models/archive/pbr-next.zip
I tried it in Chrome and Firefox for Windows, and both gave me errors.
My repository [gltf-test](https://github.com/cx20/gltf-test) has `1.3GB` in zip but was available for download.

echadwick-wayfair commented:
@emackey suggested I commit to main instead of pbr-next. My push was to pbr-next, but I used the pull request to move it to main. I'm very new to GitHub so I might have messed it up somehow.

donmccurdy commented:
Since @echadwick-wayfair's changes were written against the pbr-next branch, it will be easier to review this if the PR targets pbr-next as the base branch. I've changed that, and hopefully the GitHub diff and downloads are working now.

I'm not sure which branch this model should actually be merged into, though — is the pbr-next branch being maintained for a particular purpose? If we do want to merge this into the 'master' branch, we may need to rebase it against master or copy the changes to a new PR.

pixnblox commented:
@echadwick-wayfair: Can you add a sample image of the model here in the PR conversation? It is excellent and deserves to be seen. 👍

echadwick-wayfair commented:
OK, added a screenshot to the initial post. 

One thing I couldn't figure out was how to get refraction to work with transmission, in the yellow hanging beads or the red embedded gems. I added IOR and Volume to the Translucency version, but they don't seem to cause any visual changes in the preview in VSCode, neither Babylon.js, Cesium, Filament, nor three.js. Without visual feedback I'm not even sure if I implemented those extensions properly in the glTF.

romainguy commented:
@echadwick-wayfair IOR, volume and translucency are not implemented in Filament yet (well, we have those capabilities, just not hooked up in the glTF loader yet). They're my next priorities as soon as I can find time to write code.

bghgary commented:
Volume extension isn't merged yet in Babylon.js. Translucency and IOR should be though. Do you have a version that has these extensions?

echadwick-wayfair commented:
Babylon.js Sandbox says the version is 5.0.0-alpha.6 ...is this the latest?

When I look closely at the yellow beads, I don't see distortion of the opaque meshes, nor of the skybox. There's some blurriness from transmission roughness, but I can't see any macro-facet bending.

![2021-01-20 17_09_28-Babylon js - StainedGlassLamp_Translucency gltf](https://user-images.githubusercontent.com/40177036/105247507-e271a580-5b42-11eb-8e9e-ab0184eed661.jpg)

The yellow beads have vertex normals that are split at each vertex, essentially making each triangle a separate planar facet. I'm thinking the surface normals should affect the angles that the each light ray bends as it enters and exits the bead. Is that the way it works?

emackey commented:
Right. The volume extension is still in a branch that has a PR open for Babylon, but is not merged yet.  So there's no alpha version that will do the full refraction, only transmission.  That's what this most recent screenshot is showing.

I see you got the strings to show up, very nice!  They will look even better once Mike's Volume PR is merged.

abwood commented on 2.0/StainedGlassLamp/README.md:
```diff
@@ -0,0 +1,116 @@
+
+The model is made of eight parts, each with their own materials and using five 2048x2048 PNG textures, seven 1024x1024 PNGs, five 512x512 PNGs, and two 256x256 PNGs. All UVs are arranged between 0 and 1, except for a few overlapping parts which have been offset 1 UV unit to avoid texture baking errors. The model uses 70,588 triangles and 46,828 vertices.
+
+The model demonstrates several extensions: [KHR_materials_transmission](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_materials_transmission/README.md) for the stained glass and colored beads, [KHR_texture_clearcoat](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_materials_clearcoat/README.md) to restore shine to the surface of the transm ...
```
Minor: clearcoat extension name is incorrect.

```suggestion
The model demonstrates several extensions: [KHR_materials_transmission](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_materials_transmission/README.md) for the stained glass and colored beads, [KHR_materials_clearcoat](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_materials_clearcoat/README.md) to restore shine to the surface of the transmission-rough glass, [KHR_materials_variants](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_materials_variants/README.md) to embed two versions: "Lamp on", and "Lamp off", and [KHR_texture_basisu](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_texture_basisu/) to compress the textures. 
```

abwood commented:
I love the write up here, particularly the details provided for how this model and basisu textures were authored!

echadwick-wayfair commented:
I'm not sure why the SheenChair assets are part of this pull request. Is there a way I could split those assets off into their own pull?

emackey commented:
Hi @echadwick-wayfair,

The `pbr-next` branch is an old branch that UX3D was using to test the sample viewer, before the formation of PBR TSG.

I think you should probably copy the Stained Glass Lamp files out of this repo into a temporary folder, and then switch the repo to get onto the master branch, updated from the latest Khronos master branch.

Then, create a new branch in your repo called `stained-glass-lamp` or similar, using the latest master as a starting point.  Finally, copy the lamp files back into their new home, commit, and push the new branch up.

After that's done, you'll probably have to close this PR and open a new one from the new branch.

If you need more detail, we could move this conversation to email or zoom.  Let me know, thanks!

echadwick-wayfair commented:
Closing this branch, model has been re-uploaded in #292.
